### DISCUSSION TITLE
Large projects with multiple POM/maven files

### DISCUSSION THREAD

**Discussion by allancomar** [mfussenegger/nvim-jdtls#388] [Q&A] [answered]
Hi all,

I work with a large project that has multiple pom/maven projects that are used, for example, a common library and multiple services that consume this library.

When I code using vscode I can change the library code and use/build on the service without needing to install the package on maven. When I am using nvim I need to clean install otherwise the service is not aware of the changes.

Is there any way I can set it up so I can have the same behaviour as vscode on nvim?

Thanks.

**Comment by allancomar** [comment]
I am attaching a very simple sample. where I have a JLib that has some code and a JConsole that uses the code from the lib. if I change the code from the JLib the only way to reflect when debugging JConsole is to do a maven install on JLib.
[jproject.tar.gz](https://github.com/mfussenegger/nvim-jdtls/files/10237653/jproject.tar.gz)

**Reply by mfussenegger** [reply]
The sample project works fine for me but you need to make sure that the `root_dir` is set to the top-level `jproject` folder and not to one of the sub-folders.

E.g. if you're using something like:

```lua
root_dir = require('jdtls.setup').find_root({'pom.xml'})
```

It would detect `path/to/jproject/JConsole/` as `root_dir`, but you'll want `path/to/jproject` instead.
You'll have to adapt the `find_root` pattern to use a different marker.

**Reply by allancomar** [reply]
yes I am using a patter similar to the one above (other project types)
-- Find root of project
local root_markers = { ".git", "mvnw", "gradlew", "pom.xml", "build.gradle" }
local root_dir = require("jdtls.setup").find_root(root_markers)
if root_dir == "" then
  return
end

I am doing this on a ftplugin. Is there any way to define this per project base? meaning, there is any plugin that I could use to inject the root_dir on a project/folder, or should I try to find the pattern of which folder I am on in my ftplugin to set the root. Any advice would be great, Thanks for the help.

**Reply by allancomar** [reply]
there is any way to inspect/print the root_dir on a running instance?
Thanks.

**Reply by allancomar** [reply]
Nice. it worked with lvim. I need to figure out what I configured wrong with my LSP on nvim. but now I know is working I can find my way around it. Thanks a lot for the help.

**Reply by allancomar** [reply]
hi @mfussenegger, there is any way to print the root_dir loaded/associated with the lsp?
Thanks.

**Reply by allancomar** [reply]
guess I am finally getting to the bottom. When I do this on lunarvim I get 1 jdtls attached to the buffer on the proper folder and 1 attached to nullls
when I do on neovim I got 2 jtdls attached:
![Screen Shot 2023-06-30 at 3 29 10 PM](https://github.com/mfussenegger/nvim-jdtls/assets/13906250/1c2f6b72-9f81-4646-a087-cc2e7f18f985)

Any ideas of what I am doing wrong on the configuration? 
Thanks.

**Reply by wviana** [reply]
> hi @mfussenegger, there is any way to print the root_dir loaded/associated with the lsp?
> Thanks.

You can add print into the variable by chaing the plugin code.
Navigate into `~/.local/share/nvim/site/pack/paqs/start/nvim-jdtls` (don't know if applies to all plugin managers)
Find where the variable is defined and print it. Here is a git patch you can apply 

```diff
diff --git a/lua/jdtls/setup.lua b/lua/jdtls/setup.lua
index 07aa68a..3fb3fc8 100644
--- a/lua/jdtls/setup.lua
+++ b/lua/jdtls/setup.lua
@@ -205,6 +205,7 @@ function M.start_or_attach(config)
     or M.find_root({'.git', 'gradlew', 'mvnw'}, bufname)
     or vim.fn.getcwd()
   )
+  print(config.root_dir)
   config.handlers = config.handlers or {}
   config.handlers['language/status'] = config.handlers['language/status'] or status_callback
   config.handlers['workspace/configuration'] = config.handlers['workspace/configuration'] or configuration_handler

```


close and open nvim again, it should print it. You also can call `:messages`  to see all printed content.

**Reply by mfussenegger** [reply, marked as answer]
Or use:

`:checkhealth vim.lsp`

Or:

`:lua =vim.lsp.get_active_clients({ name = "jdtls" })` to get full details

**Comment by allancomar** [comment]
I guess I learned a little bit more about this. It seems that Microsoft has a jdtls extension to make this work on the project vscode-maven (https://github.com/microsoft/vscode-maven) that does what I am looking for. 
Does anyone know how to extend jdtls with an extension? Never tried something like that. Thanks.

**Reply by mfussenegger** [reply]
vscode-maven adds extra capabilities like searching for artifact/adding new dependencies.

It shouldn't be required to load an existing maven project. That functionality is built into eclipse.jdt.ls

**Reply by allancomar** [reply]
thanks for the clarification. I did not had much time to invest on this search on the holidays, I will for sure try the root_dir latter this week. that might resolve all my problems, I will just need to figure out a way to manage different projects. Thanks again for the help.
